File: 0018-4sum/0018-4sum.java
class Solution {
    public List<List<Integer>> fourSum(int[] nums, int target) {
        List<List<Integer>> result = new ArrayList<>();
        Arrays.sort(nums);
        for(int i=0;i < nums.length-3; i++){
            if(i-1 >=0 && nums[i] == nums[i-1]){
                continue;
            }
            
            for(int j=i+1; j < nums.length-2; j++){
                if(j-1 != i && nums[j] == nums[j-1]){
                    continue;
                }
                int left = j + 1;
                int right = nums.length - 1;
                while(left < right){
                   
                    long sum = (long) nums[i] + nums[j] + nums[left] + nums[right];
                    if(target == sum){
                        result.add(Arrays.asList(nums[i],nums[j],nums[left],nums[right]));
                        while ((left < right) && (nums[left + 1] == nums[left])) left++;
                        while ((left < right) && (nums[right - 1] == nums[right])) right--;
                        
                        left++;
                        right --;
                    } else if(sum < target){
                        
                        left ++;
                    } else {
                        right --;
                    }
                }
            }
        }
        return result;
    }
}

File: 0018-4sum/0018-4sum.py
class Solution:
    def fourSum(self, nums: List[int], target: int) -> List[List[int]]:
        result = []
        unique = set()
        nums.sort()
        for i in range(0, len(nums)-3):
            for j in range(i+1, len(nums)-2):
                start = j+1
                end = len(nums)-1
                while start < end:
                    sums = nums[i] + nums[j] + nums[start] + nums[end]
                    if sums == target:
                        unique.add(tuple([nums[i], nums[j], nums[start], nums[end]]))
                        start += 1
                        end -= 1
                    elif sums < target:
                        start += 1
                    else:
                        end -= 1
        result = list(unique)
        return result